File: remex.desktop/Services/FileTransfer/TransferProgressFormat.cs
namespace Remex.Desktop.Services.FileTransfer;

/// <summary>The unit a throughput figure is expressed in.</summary>
public enum TransferRateUnit
{
    BytesPerSecond,
    KilobytesPerSecond,
    MegabytesPerSecond,
    GigabytesPerSecond,
}

/// <summary>The unit a time-remaining figure is expressed in.</summary>
public enum TransferEtaUnit
{
    Seconds,
    Minutes,
    Hours,
}

/// <summary>A throughput figure, or the honest absence of one.</summary>
public abstract record TransferRate
{
    /// <summary>Nothing meaningful to show — no estimate yet, or an unusable one.</summary>
    public sealed record Unknown : TransferRate;

    /// <summary>A figure and the unit it is in. The caller formats and localizes it.</summary>
    public sealed record Known(double Amount, TransferRateUnit Unit) : TransferRate;
}

/// <summary>A time-remaining figure, or the honest absence of one.</summary>
public abstract record TransferEta
{
    /// <summary>No usable estimate.</summary>
    public sealed record Unknown : TransferEta;

    /// <summary>Under a second — say so rather than showing a number.</summary>
    public sealed record Finishing : TransferEta;

    /// <summary>A rounded-up amount and its unit.</summary>
    public sealed record Remaining(int Amount, TransferEtaUnit Unit) : TransferEta;
}

/// <summary>
/// Classifies throughput and time-remaining for display (RemEx-kh6p4).
/// </summary>
/// <remarks>
/// <para>
/// **PORTED AS A SHAPE, NOT AS SHARED CODE**, the same way <c>VolumesResponseClassifier</c> was
/// (RemEx-jc4q). The reference is <c>remex.android</c>'s <c>TransferProgressFormat.kt</c> and there
/// is nothing to reuse from it: different language, different resource system, different screen.
/// What transfers is the ORDER and the THRESHOLDS, and RemEx-4lcq is explicit that they should be
/// reused rather than re-derived because "none of them was obvious".
/// </para>
/// <para>
/// THIS DELIBERATELY STOPS SHORT OF WORDS. Turning <c>Remaining(3, Minutes)</c> into a sentence needs
/// plural forms, and the PC's resources have no plural infrastructure at all while Polish and
/// Ukrainian take four forms for duration words. Choosing that mechanism is a decision and it stays
/// on RemEx-4lcq; this is the half that has no decisions left in it.
/// </para>
/// <para>
/// Nothing on the PC displays speed or ETA today, so this is pre-built arithmetic rather than a
/// second definition of a rule that already runs somewhere — the distinction recorded on
/// RemEx-thwlr, and the reason it is safe to land ahead of its UI.
/// </para>
/// </remarks>
public static class TransferProgressFormat
{
    /// <summary>Beyond this, an estimate stops being information.</summary>
    /// <remarks>
    /// TWENTY-THREE HOURS RATHER THAN TWENTY-FOUR, carried over with its reasoning: a 24-hour cap
    /// makes "24 hours" the top bucket, which is a quantity nobody writes — at that size a person
    /// says "a day" — and it would sit exactly on the refusal boundary, so an estimate jittering
    /// across it would alternate between a number and nothing.
    /// </remarks>
    public const double MaximumMeaningfulSeconds = 23.0 * 60.0 * 60.0;

    private const double BytesPerKilobyte = 1024.0;
    private const double BytesPerMegabyte = 1024.0 * 1024.0;
    private const double BytesPerGigabyte = 1024.0 * 1024.0 * 1024.0;

    private const int SecondsPerMinute = 60;
    private const int MinutesPerHour = 60;
    private const int SecondsPerHour = SecondsPerMinute * MinutesPerHour;

    /// <summary>Picks the unit for a throughput figure.</summary>
    /// <param name="bytesPerSecond">Null when there is no estimate yet.</param>
    /// <remarks>
    /// NOT FINITE MEANS NOT KNOWN. This is fed from a division, and "NaN MB/s" would be a defect
    /// visible to the user in the one place they are already anxious.
    /// </remarks>
    public static TransferRate Rate(double? bytesPerSecond)
    {
        if (bytesPerSecond is not { } value || !double.IsFinite(value) || value < 0.0)
        {
            return new TransferRate.Unknown();
        }

        return value switch
        {
            >= BytesPerGigabyte => new TransferRate.Known(value / BytesPerGigabyte, TransferRateUnit.GigabytesPerSecond),
            >= BytesPerMegabyte => new TransferRate.Known(value / BytesPerMegabyte, TransferRateUnit.MegabytesPerSecond),
            >= BytesPerKilobyte => new TransferRate.Known(value / BytesPerKilobyte, TransferRateUnit.KilobytesPerSecond),
            _ => new TransferRate.Known(value, TransferRateUnit.BytesPerSecond),
        };
    }

    /// <summary>Picks the unit and the rounded amount for a time-remaining figure.</summary>
    /// <param name="secondsRemaining">Null when stalled, size unknown, or too little observed.</param>
    /// <remarks>
    /// **ROUNDED UP, THEN BUCKETED — IN THAT ORDER, AND THE ORDER IS THE WHOLE SUBTLETY.** Bucketing
    /// first gives "60 seconds left" for 59.6s, because the value is under a minute before rounding
    /// and over one afterwards. Rounding up rather than to nearest keeps the figure from ever
    /// claiming less time than remains, so the display never counts to zero and then waits — which
    /// reads as a hang.
    /// </remarks>
    public static TransferEta Eta(double? secondsRemaining)
    {
        if (secondsRemaining is not { } value || !double.IsFinite(value) || value < 0.0)
        {
            return new TransferEta.Unknown();
        }

        // Under a second there is no honest number: "0 seconds left" reads as finished, and
        // "1 second" is a rounding-up of something smaller.
        if (value < 1.0) return new TransferEta.Finishing();

        // An absurd ETA is not an ETA. A very large file at a rate just above the estimator's floor
        // still yields days, and a number that large is fiction the user would plan around.
        if (value > MaximumMeaningfulSeconds) return new TransferEta.Unknown();

        var seconds = (int)Math.Ceiling(value);
        if (seconds < SecondsPerMinute) return new TransferEta.Remaining(seconds, TransferEtaUnit.Seconds);

        var minutes = (int)Math.Ceiling(value / SecondsPerMinute);
        if (minutes < MinutesPerHour) return new TransferEta.Remaining(minutes, TransferEtaUnit.Minutes);

        return new TransferEta.Remaining((int)Math.Ceiling(value / SecondsPerHour), TransferEtaUnit.Hours);
    }
}

File: remex.android/app/src/main/java/com/clindsay94/remex/service/TransferProgressFormat.kt
package com.clindsay94.remex.service

import kotlin.math.ceil

/** Which unit a throughput figure is expressed in. */
enum class TransferRateUnit {
    BytesPerSecond,
    KilobytesPerSecond,
    MegabytesPerSecond,
    GigabytesPerSecond,
}

/** Which unit a time-remaining figure is expressed in. */
enum class TransferEtaUnit {
    Seconds,
    Minutes,
    Hours,
}

/**
 * A throughput figure, or the honest absence of one.
 *
 * [Unknown] is a first-class state rather than a zero, because "0 B/s" is a claim about the
 * transfer and "we do not know yet" is a claim about us.
 */
sealed interface TransferRate {
    data object Unknown : TransferRate

    data class Known(val value: Double, val unit: TransferRateUnit) : TransferRate
}

/**
 * A time-remaining figure.
 *
 * Three states, not two. [Finishing] exists because rendering the last fraction of a second as
 * "0 seconds left" reads as a transfer that has already completed and then visibly does not.
 */
sealed interface TransferEta {
    data object Unknown : TransferEta

    data object Finishing : TransferEta

    data class Remaining(val amount: Int, val unit: TransferEtaUnit) : TransferEta
}

/**
 * Turns [TransferRateEstimator]'s numbers into the shape a localized string needs (RemEx-qmiv).
 *
 * **STILL NOT TEXT, AND THAT IS THE POINT.** The estimator deliberately returns numbers because
 * "12.4 MB/s — 38s left" is a localized string. This layer picks the UNIT and the ROUNDING, which
 * are decisions rather than translations, and stops there. The unit symbol, the spacing, the
 * decimal separator and the plural form all belong to the resource layer, which is why nothing here
 * touches a `Context` and everything here is testable without one.
 *
 * The split matters for more than tidiness: byte-rate symbols are not universal (Ukrainian writes
 * МБ/с) and second/minute/hour words take three or four plural forms in Polish and Ukrainian, so
 * any function that returned a finished string from here would have to be wrong in at least one of
 * the nine languages this app ships in.
 */
object TransferProgressFormat {

    /**
     * Picks the unit for a throughput figure.
     *
     * @param bytesPerSecond as reported by [TransferRateEstimator.bytesPerSecond]; null means the
     *   estimator has nothing to say yet.
     */
    fun rate(bytesPerSecond: Double?): TransferRate {
        // NOT FINITE MEANS NOT KNOWN. The estimator guards its own arithmetic, but this is fed from
        // a division and a NaN rendered as "NaN MB/s" would be a defect visible to the user in the
        // one place they are already anxious.
        if (bytesPerSecond == null || !bytesPerSecond.isFinite() || bytesPerSecond < 0.0) {
            return TransferRate.Unknown
        }

        return when {
            bytesPerSecond >= BytesPerGigabyte ->
                TransferRate.Known(bytesPerSecond / BytesPerGigabyte, TransferRateUnit.GigabytesPerSecond)
            bytesPerSecond >= BytesPerMegabyte ->
                TransferRate.Known(bytesPerSecond / BytesPerMegabyte, TransferRateUnit.MegabytesPerSecond)
            bytesPerSecond >= BytesPerKilobyte ->
                TransferRate.Known(bytesPerSecond / BytesPerKilobyte, TransferRateUnit.KilobytesPerSecond)
            else -> TransferRate.Known(bytesPerSecond, TransferRateUnit.BytesPerSecond)
        }
    }

    /**
     * Picks the unit and the rounded amount for a time-remaining figure.
     *
     * @param secondsRemaining as reported by [TransferRateEstimator.secondsRemaining]; null means
     *   stalled, unknown total size, or not enough observed yet.
     */
    fun eta(secondsRemaining: Double?): TransferEta {
        if (secondsRemaining == null || !secondsRemaining.isFinite() || secondsRemaining < 0.0) {
            return TransferEta.Unknown
        }

        // Under a second there is no honest number to show: "0 seconds left" reads as finished and
        // "1 second" is a rounding-up of something smaller. Say what is actually true instead.
        if (secondsRemaining < 1.0) return TransferEta.Finishing

        // AN ABSURD ETA IS NOT AN ETA. The estimator already refuses to divide by a rate under
        // 1 KB/s, but a very large file at a rate just above that floor still yields days. A number
        // that large is fiction the user would nevertheless plan around, and the same reasoning that
        // produced the estimator's floor applies to its output.
        if (secondsRemaining > MaximumMeaningfulSeconds) return TransferEta.Unknown

        // ROUNDED UP, THEN BUCKETED - IN THAT ORDER, and the order is the whole subtlety. Bucketing
        // first gives "60 seconds left" for 59.6s, because the value is under a minute before it is
        // rounded and over one afterwards. Rounding up rather than to nearest keeps the figure from
        // ever claiming less time than remains, so the display never counts to zero and then waits.
        val seconds = ceil(secondsRemaining).toInt()
        if (seconds < SecondsPerMinute) return TransferEta.Remaining(seconds, TransferEtaUnit.Seconds)

        val minutes = ceil(secondsRemaining / SecondsPerMinute).toInt()
        if (minutes < MinutesPerHour) return TransferEta.Remaining(minutes, TransferEtaUnit.Minutes)

        return TransferEta.Remaining(
            ceil(secondsRemaining / SecondsPerHour).toInt(),
            TransferEtaUnit.Hours,
        )
    }

    /**
     * Beyond this an estimate stops being information.
     *
     * Long enough that a genuinely huge overnight transfer still reports, short enough that the
     * display never offers a figure in days off a rate barely above the estimator's stall floor.
     *
     * TWENTY-THREE HOURS RATHER THAN TWENTY-FOUR, so the largest figure the display can ever show
     * is "23 hours left". A 24-hour cap makes "24 hours" the top bucket, which is a quantity nobody
     * writes — at that size a person says "a day" — and it sits exactly on the refusal boundary, so
     * an estimate jittering across it alternates between a number and nothing.
     *
     * The flicker is not eliminated, only moved somewhere a transfer is unlikely to sit: removing it
     * entirely needs hysteresis, and that means giving a pure function memory for a case that
     * requires hovering at a specific threshold for minutes.
     */
    const val MaximumMeaningfulSeconds: Double = 23.0 * 60.0 * 60.0

    // 1024-based, matching the existing formatBytes helpers in this package. Consistency matters
    // more than the SI argument here: two figures on the same notification computed on different
    // bases would not add up.
    private const val BytesPerKilobyte = 1024.0
    private const val BytesPerMegabyte = 1024.0 * 1024.0
    private const val BytesPerGigabyte = 1024.0 * 1024.0 * 1024.0

    private const val SecondsPerMinute = 60
    private const val MinutesPerHour = 60
    private const val SecondsPerHour = 3600
}
